File: framework/languages/java/samples/java/TicTacToe.SessionGateway/Server/Session/src/main/java/systems/zlink/samples/tictactoe/sessiongateway/server/session/sessions/handlers/PlayNotificationRelay.java
package systems.zlink.samples.tictactoe.sessiongateway.server.session.sessions.handlers;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import systems.zlink.samples.tictactoe.sessiongateway.server.session.sessions.PlayerSessionDirectory;
import systems.zlink.samples.tictactoe.sessiongateway.shared.contracts.PlayResponseEnvelope;

public final class PlayNotificationRelay {
    private final PlayerSessionDirectory sessions;

    public PlayNotificationRelay(PlayerSessionDirectory sessions) {
        this.sessions = sessions;
    }

    public CompletionStage<Void> deliverAsync(String encodedResponse) {
        PlayResponseEnvelope.PlayResponse response = PlayResponseEnvelope.decode(encodedResponse);
        CompletionStage<Void> tail = CompletableFuture.completedFuture(null);
        for (PlayResponseEnvelope.Notification notification : response.notifications()) {
            tail = tail.thenCompose(ignored -> sessions
                .find(notification.recipientActorId())
                .map(context -> context.client()
                    .send(notification.payload())
                    .packetName(notification.packetName())
                    .submitAsync())
                .orElseGet(() -> CompletableFuture.completedFuture(null)));
        }
        return tail;
    }

    public String reply(String encodedResponse) {
        return PlayResponseEnvelope.decode(encodedResponse).reply();
    }
}

File: framework/languages/java/samples/kotlin/TicTacToe.SessionGateway/Server/Session/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/sessiongateway/server/session/sessions/handlers/PlayNotificationRelay.kt
package systems.zlink.samples.kotlin.tictactoe.sessiongateway.server.session.sessions.handlers

import kotlinx.coroutines.future.await
import systems.zlink.samples.kotlin.tictactoe.sessiongateway.server.session.sessions.PlayerSessionDirectory
import systems.zlink.samples.kotlin.tictactoe.sessiongateway.shared.contracts.PlayResponseEnvelope

class PlayNotificationRelay(
    private val sessions: PlayerSessionDirectory,
) {
    suspend fun deliver(encodedResponse: String) {
        val response = PlayResponseEnvelope.decode(encodedResponse)
        response.notifications.forEach { notification ->
            sessions.find(notification.recipientActorId)
                ?.client()
                ?.send(notification.payload)
                ?.packetName(notification.packetName)
                ?.submitAsync()
                ?.await()
        }
    }

    fun reply(encodedResponse: String): String =
        PlayResponseEnvelope.decode(encodedResponse).reply
}
